### PR TITLE
extmod/machine_i2c: Make the clock stretching timeout configurable and raise errors when it times out

### PR DESCRIPTION
**Pull request by deshipu**
As per discussion in #1948.

### CODE REVIEW

File: extmod/machine_i2c.c [hunk 1]
```diff
@@ -54,13 +52,16 @@ STATIC void mp_hal_i2c_scl_low(machine_i2c_obj_t *self) {
...
 
-STATIC void mp_hal_i2c_scl_release(machine_i2c_obj_t *self) {
+STATIC int mp_hal_i2c_scl_release(machine_i2c_obj_t *self) {
+    uint32_t count = self->us_timeout;
+
     mp_hal_pin_od_high(self->scl);
     mp_hal_i2c_delay(self);
     // For clock stretching, wait for the SCL pin to be released, with timeout.
-    for (int count = I2C_STRETCH_LIMIT; mp_hal_pin_read(self->scl) == 0 && count; --count) {
+    for (; mp_hal_pin_read(self->scl) == 0 && count; --count) {
```
Comment (pfalcon) [inline extmod/machine_i2c.c]: Why did you move count initialization elsewhere? Initializing it with self->us_timeout here would work just fine. The common recommendation is to minimize number of changed lines in a diff.

File: extmod/machine_i2c.c [hunk 2]
```diff
@@ -54,13 +52,16 @@ STATIC void mp_hal_i2c_scl_low(machine_i2c_obj_t *self) {
...
         mp_hal_delay_us_fast(1);
     }
+    return !!count;
```
Comment (pfalcon) [inline extmod/machine_i2c.c]: This is a known C idiom, but not commonly used in uPy. @dpgeorge prefers proper boolean expressions, like "count != 0". And if it really returns bool, then function type should apparently be that too.

File: extmod/machine_i2c.c (hunk 1 above)
Comment (pfalcon) [inline extmod/machine_i2c.c]: D'oh, nevermind, removing this comment ;-)

File: extmod/machine_i2c.c (hunk 2 above)
Comment (deshipu) [inline extmod/machine_i2c.c]: I will replace it with `count != 0` then. As for using `bool`, that's what I initially did, but then I noticed all the other functions in this file and others that return success/failure and use `int`, and decided to make it consistent.

File: extmod/machine_i2c.c [hunk 3]
```diff
@@ -75,21 +76,23 @@ STATIC int mp_hal_i2c_sda_read(machine_i2c_obj_t *self) {
...
-STATIC void mp_hal_i2c_start(machine_i2c_obj_t *self) {
+STATIC int mp_hal_i2c_start(machine_i2c_obj_t *self) {
     mp_hal_i2c_sda_release(self);
     mp_hal_i2c_delay(self);
-    mp_hal_i2c_scl_release(self);
+    int ret = mp_hal_i2c_scl_release(self);
     mp_hal_i2c_sda_low(self);
     mp_hal_i2c_delay(self);
+    return ret;
 }
```
Comment (dpgeorge) [inline extmod/machine_i2c.c]: If there is a bus error, don't you think the I2C driver should release both SCL and SDA lines before returning?  In this function SCL is released, and so is SDA, right after the call to mp_hal_i2c_scl_release, so you can simply return at that point if the call fails.

File: extmod/machine_i2c.c [hunk 4]
```diff
@@ -113,13 +116,17 @@ STATIC int mp_hal_i2c_write_byte(machine_i2c_obj_t *self, uint8_t val) {
             mp_hal_i2c_sda_low(self);
         }
         mp_hal_i2c_delay(self);
-        mp_hal_i2c_scl_release(self);
+        if (!mp_hal_i2c_scl_release(self)) {
+            return 0; // failure
```
Comment (dpgeorge) [inline extmod/machine_i2c.c]: I guess here you want to do mp_hal_i2c_sda_release()?

**Review by dpgeorge — COMMENTED** [review]
It's a decision we have to make, whether the specs of our SW I2C driver say that both lines are released if there is a bus error.

**Comment by deshipu** [comment]
Sorry, somehow I missed the comments here.

**Comment by dpgeorge** [comment]
Thanks @deshipu.  Rebased and merged, see e81a5353cb794b8d6c57317488e5db6a1c524755.
